pyacq/devices/audio_pyaudio.py

```diff
@@ -0,0 +1,156 @@
+                    rate=int(self.sampling_rate), input_device=self.input_device_index),\
+                    'Input not supported {} {} device {}'.format(self.nchannel, samplerate, self.input_device_index)
+        
+        if self.output_device_index is not None:
+            assert self.pa.is_format_supported(output_format=format_conv[format], output_channels=self.nb_channel, 
+                    rate=int(self.sampling_rate), output_device=self.output_device_index),\
+                    'Output not supported {} {} device {}'.format(self.nchannel, samplerate, self.input_device_index)
+
+        self.output.spec['shape'] = (self.chunksize, self.nb_channel)
+        self.output.spec['dtype = '] = format
+        self.output.spec['sampling_rate'] = float(int(self.sampling_rate))
+        gains = { 'int16' : 1./2**15, 'int32' : 1./2**31, 'float32':1. }
+        self.output.spec['gain'] = gains[self.format]
+        self.output.spec['offset'] = 0.
+    
+    def check_input_specs(self):
+        pass
+    
+    def check_output_specs(self):
+        pass
+
+    def _initialize(self):
+        self.audiostream = self.pa.open(
+                    rate = int(self.sampling_rate),
+                    channels = int(self.nb_channel),
+                    format = format_conv[self.format],
+                    input= self.input_device_index is not None,
+                    output= self.output_device_index is not None,
+                    input_device_index = self.input_device_index,
+                    output_device_index = self.output_device_index,
+                    frames_per_buffer = self.chunksize,
+                    stream_callback = self._audiocallback,
+                    start = False)
+        self.head = 0
+        
+        # outbuffer
+        size = self.nb_channel * self.chunksize * np.dtype(self.format).itemsize
+        self.enmpty_outputbuffer = b'\x00' *  size
+        self.out_queue = collections.deque()
+        self.lock = Mutex()
+        
+        if self.output_device_index is not None:
+            self.thread = ThreadPollInput(self.input, parent = self)
+            self.thread.new_data.connect(self._new_output_buffer)
+    
+    def _start(self):
+        self.audiostream.start_stream()
+        if self.output_device_index is not None:
+            self.thread.start()
+
+    def _stop(self):
+        self.audiostream.stop_stream()
+        if self.output_device_index is not None:
+            self.thread.stop()
+            self.thread.wait()
+
+    def _close(self):
+        self.audiostream.close()
+        self.pa.terminate()
+        del self.thread
+
+    def _audiocallback(self, in_data, frame_count, time_info, status):
+        #~ print('audiocallback', len(self.out_queue))
+        if in_data is not None:
+            self.head += self.chunksize
+            self.output.send(self.head, in_data)
```

Comment:
I think there is a field in between your suspiciousness thread approach and my over confidence thread approach.